extensions/wikia/TemplateClassification/maintenance/classifyNonArticleTemplates.class.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+$mainDir = __DIR__ . "/../../../..";
+require_once( $mainDir . '/maintenance/Maintenance.php' );
+
+class ClassifyNonArticleTemplates extends Maintenance {
+
+	const NONARTICLE_MAINTENANCE_PROVIDER = 'maintenance';
+	const NONARTICLE_MAINTENANCE_ORIGIN = 1; // Since it's the first script of this type
+
+	private $dryRun, $quiet, $logFile;
+
+	public function __construct() {
+		parent::__construct();
+
+		$this->addOption( 'dry-run', 'Set if you just want to count first-level and nested templates.' );
+		$this->addOption( 'quiet', 'Set if you do not want to output any logs.' );
+		$this->addOption( 'log-file', 'If you need logs in a .csv format provide a path to an existing file.', false, true );
+	}
+
+	public function execute() {
+		global $wgCityId, $wgDBname, $wgContentNamespaces;
+
+		$this->dryRun = $this->getOption( 'dry-run' );
+		$this->quiet = $this->getOption( 'quiet' );
+		$this->logFile = $this->getOption( 'log-file' );
+
+		if ( isset( $this->logFile ) && !$this->verifyLogFile() ) {
+			$this->out( "The log file {$this->logFile} does not exist, please create it." );
+			return false;
+		}
+
+		$this->setDB( wfGetDB( DB_SLAVE ) );
+		$tcs = new TemplateClassificationService();
+
+		$namespacesTemplates = ( new \WikiaSQL() )
+			->SELECT( 'pt.page_title as template_title, pt.page_id as template_id, p.page_id as page_id' )
+			->FROM( 'templatelinks' )->AS_( 'tl' )
+			->INNER_JOIN( 'page' )->AS_( 'pt' )
+			->ON( 'tl.tl_title', 'pt.page_title' )
+			->INNER_JOIN( 'page' )->AS_( 'p' )
+			->ON( 'tl.tl_from', 'p.page_id' )
+			->WHERE( 'tl.tl_namespace' )->EQUAL_TO( NS_TEMPLATE )
+			->AND_( 'pt.page_namespace' )->EQUAL_TO( NS_TEMPLATE )
+			->AND_( 'p.page_namespace' )->IN( $wgContentNamespaces )
+			->runLoop( $this->getDB(), function ( &$pages, $row ) {
+				if ( !isset( $pages[$row->template_id]['title'] ) ) {
+					// First run for this template ID
+					$pages[$row->template_id]['title'] = $row->template_title;
+					$pages[$row->template_id]['linkingPages'] = [];
+				}
+
+				$pages[$row->template_id]['linkingPages'][] = (int)$row->page_id;
+			} );
+
+		$countFirstLevelTemplates = 0;
+		foreach ( $namespacesTemplates as $templateId => $data ) {
+			$templateTitle = $data['title'];
+			$linkingPages = $data['linkingPages'];
+			$count = count( $linkingPages );
+
+			$this->out( "\nProcessing {$count} inclusions of template {$templateTitle} ({$templateId})" );
+			$isFirstLevel = false;
+
+			foreach ( $linkingPages as $pageId ) {
+				$pageRawTextLc = strtolower( WikiPage::newFromID( $pageId )->getRawText() );
```

Comment:
It's rare, but we have had stale data with no longer existent pages stuck in templatelinks, categorylinks, and similar tables before. Just as a sanity check, I'd consider making sure `WikiPage::newFromID( $pageId )` returns a `WikiPage` object as if the page ID can't be found, it will return null and this will fatal.